finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala

```diff
@@ -442,8 +442,10 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
 
           case Open(remote@RemoteStreaming()) =>
             if (stateRef.compareAndSet(state, remote.toRemoteClosed)) {
-              if (recvFrame(toFrame(data), remote)) true
-              else throw new IllegalStateException("stream queue closed prematurely")
+              if (recvFrame(toFrame(data), remote)) {
+                remote.close()
+                true
+              } else throw new IllegalStateException("stream queue closed prematurely")
```

Comment:
If `recvFrame` returns false, that means that the call to `remote.offer` failed, due to the stream already being closed. So I think in this case calling `remote.close` again would be redundant.